File: Node/Euler018.js
#!/usr/bin/env node
const euler011 = require('./Euler011.js');

function main() {
  console.log(getTriangleBiggestSum("input/Euler018.txt"));
}

function getTriangleBiggestSum(ofFile) {
  triangle = euler011.loadFile(ofFile);

  for (var row = triangle.length - 2; row >= 0; row--) {
    // calculating subtriangles. The bigger option of each "edge" in a subtriangle is the optimal route.
    // we work by modifying each element to be the max of its available paths.
    for (var col = 0; col < triangle[row].length; col++) {
      let lchild = triangle[row + 1][col];
      let rchild = triangle[row + 1][col + 1];

      if (lchild > rchild) {
        triangle[row][col] += lchild;
      } else {
        triangle[row][col] += rchild;
      }
    }
  }

  return triangle[0][0];
}
exports.getTriangleBiggestSum = getTriangleBiggestSum;

if (require.main == module) {
  main();
}

File: Python/Euler018.py
#!/usr/bin/env python3
import Euler011


def main():
    print(get_biggest_triangle_sum('input/Euler018.txt'))


def get_biggest_triangle_sum(of_file):
    triangle = Euler011.load_file(of_file)

    for row in reversed(range(0, len(triangle) - 1)):
        for col in range(0, len(triangle[row])):
            lchild = triangle[row+1][col]
            rchild = triangle[row+1][col+1]

            if lchild > rchild:
                triangle[row][col] += lchild
            else:
                triangle[row][col] += rchild

    return triangle[0][0]


if __name__ == '__main__':
    main()
